### PR TITLE
Fix numerical diffusion in uniformMesh converter

### PR DESCRIPTION
**Pull request by mgjarrett**
## Description

Mapping anything from a non-uniform assembly to a uniform assembly and then back will cause numerical diffusion of that value. This could be a block parameter (like mgFlux, detailedDpa, etc.) or number densities. This has been observed for neutron flux while using the `nonUniformAssemFlags` setting on control rods, and for `detailedDpa` when using the `detailedAxialExpansion` setting.

This commit deals with the issue of block parameters being mapped back and forth. The number density issue has already been addressed for the nonUniformAssemFlags case, but it seems to still be present for the detailedAxialExpansion case.

The selection of which parameters to map is very much up for discussion. **Please review the categories that have been set for `REACTOR_PARAMS_TO_MAP` and `BLOCK_PARAMS_TO_MAP`, and the `_setParamsToUpdate` method for both neutron and gamma transport.** Please leave comments on any parameters you think are missing or should be changed. My approach to these selections is summarized below:

- First and foremost, we want to ensure that we are not mapping any parameters both "into" and "out of" the uniform mesh assembly, because that will cause numerical diffusion.
- We also want to make sure that when a new parameter value is generated on the uniform mesh by some physics solver, that parameter is mapped back to the non-uniform mesh.
- We should always map number densities "into" the uniform assembly but never out of it. Other parameters (like power, temperature, etc.) generally don't need to be mapped in unless they are required to set up the input for the physics kernel. It will generally be harmless to include these additional parameters **so long as they are not re-mapped back to the non-uniform assembly on the way out.**

---

## Checklist

- [x] This PR has only one purpose or idea.
- [x] Tests have been added/updated to verify that the new/changed code works.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any bug fixes or new features.
- [x] The documentation is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `setup.py`.

### CODE REVIEW

**Comment by mgjarrett** [comment]
Not requesting review yet, just tagging a few interested parties @keckler @Nebbychadnezzar

**Comment by mgjarrett** [comment]
@Nebbychadnezzar can't select you as a reviewer but I wanted to get your eyes on this as well.

File: armi/reactor/converters/uniformMesh.py
Comment (keckler) [inline armi/reactor/converters/uniformMesh.py:546]: My understanding of this change is as follows:

- Before, we were clearing all of the block parameters on the new, converted assembly, while storing some of them temporarily in a "cache". Then, we were mapping the specified parameters to the new assembly, and the cached parameters are then used to put the other parameters back onto the blocks.
- Now, we just avoid clearing the parameters in the first place, so that the caching of parameters isn't needed for anything. All parameters that are not specified specifically to be mapped are just left as their initial values?

This seems like potentially a source of great speedup. But it makes me wonder, are we losing something by changing that flow? It seems to me that an expensive step, like clearing all the parameters in the first place, must have been done for a good reason. But I cannot see why that reason might have gone away now...

File: armi/reactor/converters/uniformMesh.py
Comment (keckler) [inline armi/reactor/converters/uniformMesh.py:618]: Should this method instead be changed to raise a `NotImplemetedError`, since we never want people to actually instantiate this base converter class?

I'm not sure, your call.

File: armi/reactor/converters/uniformMesh.py
Comment (keckler) [inline armi/reactor/converters/uniformMesh.py:782]: This call doesn't do anything, except set up some empty lists. Which is fine. But if you wanted to go the `NotImplementedError` route that I suggested in another comment, then this obviously won't work.

File: armi/reactor/converters/uniformMesh.py
Comment (keckler) [inline armi/reactor/converters/uniformMesh.py:857]: I would move this comment to the docstring as a `Note`. It's useful for people to see.

**Review by keckler — APPROVED** [review]
I like it. I left a handful of comments and suggestions, but I don't see anything blatantly wrong. 

After this is merged, though, there is still work to do in our internal plugins to move the application/calculation of cumulative parameters outside of the geometry conversion entry/exit. For instance, I can say for certain that `detailedDpaRate` is calculated right now when the geometry is still converted. Likely the same for many of the `cumulative` parameters. So we'll have to hunt all of those down and update them, if necessary. 

For `cumulative` parameters that are being calculated within the framework itself, I would argue it might make sense to update those as part of this PR (if any of them exist).

File: armi/reactor/parameters/parameterDefinitions.py
Comment (mgjarrett) [reply armi/reactor/parameters/parameterDefinitions.py:70]: Yes, it is confusing. If only neutron transport is run without gamma, then the `power` and `pdens` parameters are populated. Once gamma is run, then `powerGamma` and `powerNeutron` are populated.

I think they have to stay in this category, so maybe I'll include a comment or update the parameter description.

File: armi/reactor/blocks.py
Comment (mgjarrett) [reply armi/reactor/blocks.py:32]: This line looks like it's carried over from some experimental stuff. I was originally trying to do something more complicated with creating new blocks in `makeAssemWithUniformMesh` but I gave up on that for now.

File: armi/reactor/converters/uniformMesh.py
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:179]: Ah, this actually needs to be changed back.

`_buildAllUniformAssemblies()` calls `makeAssemWithUniformMesh()`. By default, `makeAssemWithUniformMesh()` was previously doing `mapNumberDensities=True`. I changed the default to `False` at some point during the work on this PR and then probably set this to `True` here so the number densities will still be mapped. But changing the default of `makeAssemWithUniformMesh()` turned out to be a mistake so it is now back to the original default. With the original default, `_buildAllUniformAssemblies()` takes care of number densities and this should be set to `False`.

File: armi/reactor/converters/uniformMesh.py
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:279]: The number densities are updated using `setNumberDensitiesFromOverlaps` which calls some specific functions that update number densities on all the components. The logic for updating number densities is a little different than the other parameters, so it probably makes sense to keep them separate.

File: armi/reactor/converters/uniformMesh.py
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:437]: Yeah the old description is better. I'm pretty sure what happened is that at some point I had deleted `blockParamNames` entirely, and then brought it back. So the less descriptive version was created from scratch, not an attempt to improve the original description.

File: armi/reactor/converters/uniformMesh.py
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:618]: That's a good point. But theoretically if you're only interested in geometry and not parameters, it might be valid to instantiate the base class and use it?

File: armi/reactor/converters/uniformMesh.py
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:782]: I initially thought there might be some parameters that we want to map in all cases that would be defined in the base class implementation. But I can't think of any good examples, so... ?

File: armi/reactor/converters/uniformMesh.py
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:546]: My understanding of the caching is basically the same as yours.

As for whether the caching is necessary, I would break it down into the two directions:

- Non-uniform source, uniform destination assembly: When the destination assembly is the uniform assembly, it's being created from the source assembly, so it shouldn't have any pre-existing parameters that need to be cached.
- Uniform source, non-uniform destination: When the destination assembly is the non-uniform assembly, parameter caching would ensure that we don't erase any of the parameters in `blockParamNames` on the non-uniform assembly by overwriting with null values from the source assembly. But, if we have defined the `blockParamNames` to be mapped back correctly, then we should only be mapping back parameters that have been set on the uniform assembly. If this `setAssemblyStateFromOverlaps` is trying to map back parameters that haven't been set, then either the `_setParamsToUpdate` was not defined correctly or maybe some analysis performed on the uniform reactor failed. `_applyCachedParamValues` would paper over that error by just retaining whatever value the non-uniform assembly already had for that parameter, but this would create a "silent" bug that would be very difficult to track down (e.g., you have a `b.p.power` in the database, but you have no way of verifying whether it was mapped back correctly or just retained by caching).

That's my take on it, but it's complicated and I'm definitely curious to hear the opinions of others. I think the short summary is that parameter caching would not be performing any detectable function when everything is working correctly. It would only act when there is a bug in the code or an error during execution, and it would effectively cover up the error without giving any warning or error message.

**Comment by mgjarrett** [comment]
> After this is merged, though, there is still work to do in our internal plugins to move the application/calculation of cumulative parameters outside of the geometry conversion entry/exit. For instance, I can say for certain that `detailedDpaRate` is calculated right now when the geometry is still converted. Likely the same for many of the `cumulative` parameters. So we'll have to hunt all of those down and update them, if necessary.

I am aware of `detailedDpa`, and that looks like it's going to somewhat difficult to address. I don't know of anything else in the downstream plugins, though.

> For `cumulative` parameters that are being calculated within the framework itself, I would argue it might make sense to update those as part of this PR (if any of them exist).

I don't know of any parameter accumulation within the framework itself that needs to be updated. The most obvious cumulative parameter is depletion/burnup, and those are calculated on the non-uniform assembly. Aside from depletion/residence/burnup, fluence, DPA (mentioned above), I can't think of any other obvious cumulative parameters.

File: armi/reactor/converters/uniformMesh.py
Comment (keckler) [inline armi/reactor/converters/uniformMesh.py:741]: I am wondering why the name of this dict was changed? To me, the original name `BLOCK_PARAM_MAPPING_CATEGORIES` was more descriptive of what is actually contained in the dict.

The new name `BLOCK_PARAMS_TO_MAP` indicates to me that it contains parameter names, but this is not the case.

File: armi/reactor/converters/uniformMesh.py
Comment (keckler) [inline armi/reactor/converters/uniformMesh.py:734]: I think this note might be out of date? Or maybe just flat-out wrong? Mapping the parameters from and then back to a source reactor is exactly what we do _not_ want to do. 

Does this need to be updated?

File: armi/reactor/converters/uniformMesh.py
Comment (keckler) [inline armi/reactor/converters/uniformMesh.py:844]: I think it would be extremely useful to outline in the docstring for this class as well as the `NeutronicsUniformMeshConverter` the conditions which determine which parameters are mapped in each direction. 

I foresee a case where a plugin is mistakenly calculating a parameter that is not mapped back to the source reactor, and then the user/developer would be left wondering why their updated value did not get reflected back on their original reactor.

Making it clear in the docstring that certain parameters are not mapped to/from would help to guard against that kind of situation.

**Comment by keckler** [comment]
> I don't know of any parameter accumulation within the framework itself that needs to be updated. The most obvious cumulative parameter is depletion/burnup, and those are calculated on the non-uniform assembly. Aside from depletion/residence/burnup, fluence, DPA (mentioned above), I can't think of any other obvious cumulative parameters.

I agree with this sentiment. I just scanned through the framework parameters and couldn't find any which are cumulative that would also be mapped in the first place. There are some things in the `fuelPerformance` package that are cumulative (gasReleaseFraction, bondRemoved, cladWastage, totalCladStrain), but I guess those won't be mapped in the first place because they don't satisfy the category qualifications.

So we're probably good there!

**Comment by keckler** [comment]
I pulled this PR into my ARMI app, reran some extensive calculations, and then looked at the calculated DPAs. 

I can confirm that this PR resolves the issues I was previously having with DPA possibly decreasing as time increases. Obviously that was non-physical, and these changes resolve that such that we are now getting physically-possible results for DPA.

I can also add another data point to add confidence to this change. The model that I ran is a benchmark of real reactor operations, and I was previously getting anomalously low reactivities in later cycles. I had suspected that this was due to the numerical diffusion of the uniform mesh generator, and it turns out that with this fix, core reactivities line up much more closely than before. 

So in my book, the changes in this PR are a resounding success. I approve!

**Comment by mgjarrett** [comment]
#985

File: armi/reactor/parameters/parameterDefinitions.py
Comment (john-science) [reply armi/reactor/parameters/parameterDefinitions.py:70]: I would love a docstring or description in the /docs/ somewhere that explains to people what to use these two new categories for.

I mean, _I_ see your description above about going into/out of the uniform mesh.  But we want people who come to ARMI in a year to be able to understand this, right?

File: armi/reactor/converters/uniformMesh.py
Comment (john-science) [reply armi/reactor/converters/uniformMesh.py:546]: My major concern with preserving the old, caching solution is backwards compatibility for our downstream projects.

But it doesn't appear well enough documented that I'm thrilled with having to support it forever. Michael's above description matches what I see in the code. But the original developer could have saved us several hours of digging if it had been better documented.

Let's do better!

**Comment by john-science** [comment]
> I pulled this PR into my ARMI app, reran some extensive calculations, and then looked at the calculated DPAs.

So... above you were concerned that "there is still work to do in our internal plugins to move the application/calculation of cumulative parameters outside of the geometry conversion entry/exit".  

Are you concerns with that over?

File: armi/reactor/converters/tests/test_uniformMesh.py
Comment (john-science) [inline armi/reactor/converters/tests/test_uniformMesh.py:44]: Can we remove this `print()` statement?  We try not to sprinkle `print`s throughout ARMI.

Comment (mgjarrett) [reply armi/reactor/converters/tests/test_uniformMesh.py:44]: ah yes, good catch

**Comment by keckler** [comment]
> Are you concerns with that over?

Yes, I am sufficiently convinced after both looking though our internal code and testing this PR that we are in the clear on that concern.

File: armi/reactor/parameters/parameterDefinitions.py
Comment (mgjarrett) [reply armi/reactor/parameters/parameterDefinitions.py:82]: I added some docstrings to describe most of the parameter categories in this class.

File: armi/reactor/converters/uniformMesh.py
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:773]: I never saw this comment somehow.

No good reason for it other than the order in which development happened. I started by trying to create a list of parameter names, not categories, so I changed the attribute name. And then I went back to categories.

Another good question is why is this snake case? I have no idea. I should probably change it to `blockParamMappingCategories`.

File: armi/reactor/converters/uniformMesh.py
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py]: Yes it was just the old doc string, I had not updated it

File: armi/reactor/converters/uniformMesh.py
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:905]: Somehow I missed several of these comments the first time I looked through.

I added a lot of detail to the base class `UniformMeshGeometryConverter` docstring and to the derived class docstrings. I could definitely see that being a really frustrating error to contend with, so I'll try to put a clear warning in the docstring of each converter.

**Comment by LMikeH** [comment]
I'm not sure why, but I'm getting an error in JOYO that I thought I fixed previously. I looked to make sure my previous changes were still there and they are. I don't know if any of you have seen this error before:

  File "\\albert\home\mhuang\site-packages\nala\__main__.py", line 26, in main
    code = ArmiCLI().run()
  File "\\albert\home\mhuang\site-packages\framework\armi\cli\__init__.py", line 202, in run
    return self.executeCommand(args.command, args.args)
  File "\\albert\home\mhuang\site-packages\framework\armi\cli\__init__.py", line 231, in executeCommand
    return cmd.invoke()
  File "\\albert\home\mhuang\site-packages\framework\armi\cli\run.py", line 33, in invoke
    inputCase.run()
  File "\\albert\home\mhuang\site-packages\framework\armi\cases\case.py", line 367, in run
    o.operate()
  File "\\albert\home\mhuang\site-packages\framework\armi\operators\operatorMPI.py", line 74, in operate
    Operator.operate(self)
  File "\\albert\home\mhuang\site-packages\framework\armi\operators\operator.py", line 320, in operate
    self._mainOperate()
  File "\\albert\home\mhuang\site-packages\framework\armi\operators\operator.py", line 329, in _mainOperate
    keepGoing = self._cycleLoop(cycle, startingCycle)
  File "\\albert\home\mhuang\site-packages\framework\armi\operators\operator.py", line 345, in _cycleLoop
    halt = self.interactAllBOC(self.r.p.cycle)
  File "\\albert\home\mhuang\site-packages\framework\armi\operators\operator.py", line 552, in interactAllBOC
    return self._interactAll("BOC", activeInterfaces, cycle)
  File "\\albert\home\mhuang\site-packages\framework\armi\operators\operator.py", line 423, in _interactAll
    halt = halt or interactMethod(*args)
  File "\\albert\home\mhuang\site-packages\framework\armi\utils\codeTiming.py", line 55, in time_wrapper
    return_value = func(*args, **kwargs)
  File "\\albert\home\mhuang\site-packages\framework\armi\physics\neutronics\latticePhysics\latticePhysicsInterface.py", line 108, in interactBOC
    self.updateXSLibrary(cycle)
  File "\\albert\home\mhuang\site-packages\framework\armi\physics\neutronics\latticePhysics\latticePhysicsInterface.py", line 129, in updateXSLibrary
    blockList=representativeBlocks, xsLibrarySuffix=self._getSuffix(cycle)
  File "\\albert\home\mhuang\site-packages\plugins\terrapower-partisn\terrapower\physics\neutronics\partisn\mc2PartisnInterface.py", line 78, in computeCrossSections
    executer.run()
  File "\\albert\home\mhuang\site-packages\plugins\terrapower-partisn\terrapower\physics\neutronics\partisn\partisnExecuters.py", line 417, in run
    output = globalFluxInterface.GlobalFluxExecuter.run(self)
  File "\\albert\home\mhuang\site-packages\framework\armi\physics\executers.py", line 187, in run
    self._performGeometryTransformations()
  File "\\albert\home\mhuang\site-packages\plugins\terrapower-partisn\terrapower\physics\neutronics\partisn\mc2PartisnExecuters.py", line 88, in _performGeometryTransformations
    self._computeFineGroupXS()
  File "\\albert\home\mhuang\site-packages\plugins\terrapower-partisn\terrapower\physics\neutronics\partisn\mc2PartisnExecuters.py", line 99, in _computeFineGroupXS
    blockList=self._xsgm.representativeBlocks.values()
  File "\\albert\home\mhuang\site-packages\plugins\terrapower-partisn\terrapower\physics\neutronics\partisn\mc2PartisnInterface.py", line 117, in computeFineGroupXs
    purgeFP=False,
  File "\\albert\home\mhuang\site-packages\plugins\terrapower-mc2\terrapower\physics\neutronics\mc2\mc2Interface.py", line 236, in _generateXsLibrary
    baseList, xsLibrarySuffix, blockList, writers
  File "\\albert\home\mhuang\site-packages\framework\armi\physics\neutronics\latticePhysics\latticePhysicsInterface.py", line 300, in generateLatticePhysicsInputs
    fromWriter = writer.write()
  File "\\albert\home\mhuang\site-packages\plugins\terrapower-mc2\terrapower\physics\neutronics\mc2\mc2Writers.py", line 411, in write
    self._writeComposition(f)
  File "\\albert\home\mhuang\site-packages\plugins\terrapower-mc2\terrapower\physics\neutronics\mc2\mc2Writers.py", line 601, in _writeComposition
    self._writeSingleComposition(fileObj, compNumber, component)
  File "\\albert\home\mhuang\site-packages\plugins\terrapower-mc2\terrapower\physics\neutronics\mc2\mc2Writers.py", line 606, in _writeSingleComposition
    allNuclides = self._getAllNuclideObjects(component)
  File "\\albert\home\mhuang\site-packages\framework\armi\physics\neutronics\latticePhysics\latticePhysicsWriter.py", line 187, in _getAllNuclideObjects
    nucs, fissProds = self._getAllNuclidesByCategory(component)
  File "\\albert\home\mhuang\site-packages\framework\armi\physics\neutronics\latticePhysics\latticePhysicsWriter.py", line 204, in _getAllNuclidesByCategory
    dfpDensities = self._getDetailedFPDensities()
  File "\\albert\home\mhuang\site-packages\framework\armi\physics\neutronics\latticePhysics\latticePhysicsWriter.py", line 370, in _getDetailedFPDensities
    dfpDensitiesByName = lfpCollection.getNumberDensities(self.block)
AttributeError: 'NoneType' object has no attribute 'getNumberDensities'

**Comment by keckler** [comment]
@LMikeH I'll message you offline.

**Review by john-science — APPROVED** [review]
I think this is a good solution, in that it solves a semi-complicated problem.

The solution is as complicated as the problem.  But sometimes there is just no way around that.

As long as we are all convinced by the documentation we've put in place to explain it to future generations, I'm happy.

**Pull request merged**
